File: windows/Tokendial.Core/Alerts/AlertEngine.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Tokendial.Core.Alerts;

/// <summary>
/// The deterministic reducer specified in docs/alerts/spec.md. It owns no
/// clock and no thread: the host feeds it events and delivers what comes back.
/// Every (provider, window) lives in an epoch keyed by its reset time; a new
/// epoch re-arms everything.
/// </summary>
public sealed class AlertEngine
{
    private static readonly JsonSerializerOptions Json = new()
    {
        WriteIndented = true,
        Converters = { new JsonStringEnumConverter() }
    };

    private readonly AlertConfig config;
    private readonly AlertState state;

    public AlertEngine(AlertConfig? config = null, AlertState? state = null)
    {
        this.config = config ?? AlertConfig.Default;
        this.state = state ?? new AlertState();
    }

    public static AlertEngine Load(string json, AlertConfig? config = null)
    {
        AlertState? state = null;
        try
        {
            state = JsonSerializer.Deserialize<AlertState>(json, Json);
        }
        catch (JsonException)
        {
        }
        return new AlertEngine(config, state?.Schema == 1 ? state : null);
    }

    public string Save() => JsonSerializer.Serialize(state, Json);

    public IReadOnlyList<Alert> Reduce(AlertEvent e)
    {
        var candidates = new List<Alert>();
        switch (e)
        {
            case AlertEvent.Usage usage:
                candidates.AddRange(ApplyUsage(usage));
                break;
            case AlertEvent.Session session:
                ApplySession(session);
                break;
            case AlertEvent.Hover hover:
                state.HoverOn = hover.On;
                break;
            case AlertEvent.Restart restart:
                Prune(restart.At);
                break;
        }
        candidates.AddRange(TimeChecks(e.At));
        return Gate(candidates, e.At);
    }

    private IEnumerable<Alert> ApplyUsage(AlertEvent.Usage usage)
    {
        var key = $"{usage.Provider}|{usage.Window}";
        var epoch = state.Epochs.FirstOrDefault(x => x.Key == key);
        var fresh = epoch is null || epoch.Closed || epoch.ResetsAt != usage.ResetsAt || epoch.LastPct - usage.UsedPct > 10;
        if (fresh)
        {
            state.Epochs.RemoveAll(x => x.Key == key);
            epoch = new Epoch { Key = key, Provider = usage.Provider, Window = usage.Window, ResetsAt = usage.ResetsAt };
            state.Epochs.Add(epoch);
        }
        epoch!.LastPct = usage.UsedPct;
        epoch.ResetsAt = usage.ResetsAt;

        if (usage.Limited || usage.UsedPct >= 100)
        {
            foreach (var threshold in config.Thresholds) epoch.Fired.Add(threshold.ToString());
            if (epoch.Fired.Add("limit"))
            {
                yield return new Alert(AlertKind.Limit, usage.Provider, usage.Window, ResetsAt: usage.ResetsAt);
            }
            yield break;
        }

        var crossed = config.Thresholds.Where(t => t <= usage.UsedPct && !epoch.Fired.Contains(t.ToString())).ToList();
        if (crossed.Count == 0) yield break;
        foreach (var threshold in crossed) epoch.Fired.Add(threshold.ToString());
        yield return new Alert(AlertKind.Threshold, usage.Provider, usage.Window, Pct: crossed.Max());
    }

    private void ApplySession(AlertEvent.Session session)
    {
        var key = $"{session.Provider}|{session.SessionId}";
        if (session.State == SessionActivity.Waiting)
        {
            if (state.Waiting.All(w => w.Key != key))
            {
                state.Waiting.Add(new WaitingEntry { Key = key, Provider = session.Provider, SessionId = session.SessionId, Since = session.At });
            }
            return;
        }
        state.Waiting.RemoveAll(w => w.Key == key);
    }

    private IEnumerable<Alert> TimeChecks(DateTimeOffset now)
    {
        foreach (var epoch in state.Epochs.Where(x => !x.Closed && x.ResetsAt is not null).ToList())
        {
            var resetsAt = epoch.ResetsAt!.Value;
            if (!epoch.ResetSoonFired && !epoch.Fired.Contains("limit") && resetsAt > now
                && resetsAt - now <= config.ResetLead && epoch.LastPct >= config.ResetLeadMinPct)
            {
                epoch.ResetSoonFired = true;
                yield return new Alert(AlertKind.ResetSoon, epoch.Provider, epoch.Window, ResetsAt: resetsAt);
            }
            if (now >= resetsAt && epoch.Fired.Contains("limit") && !epoch.ResetDoneFired)
            {
                epoch.ResetDoneFired = true;
                epoch.Closed = true;
                yield return new Alert(AlertKind.ResetDone, epoch.Provider, epoch.Window);
            }
        }

        foreach (var waiting in state.Waiting)
        {
            var due = waiting.LastFired is DateTimeOffset last ? last + config.WaitingRepeat : waiting.Since + config.WaitingDebounce;
            if (now < due) continue;
            waiting.LastFired = now;
            yield return new Alert(AlertKind.Waiting, waiting.Provider, SessionId: waiting.SessionId);
        }
    }

    /// <summary>Hover and cooldown decide what goes out now, what waits, and what is dropped.</summary>
    private IReadOnlyList<Alert> Gate(List<Alert> candidates, DateTimeOffset now)
    {
        var emitted = new List<Alert>();

        if (!state.HoverOn)
        {
            foreach (var held in state.Held.ToList())
            {
                if (InCooldown(held.Provider, now)) continue;
                state.Held.Remove(held);
                emitted.Add(held);
            }
        }

        foreach (var alert in candidates)
        {
            if (alert.Kind == AlertKind.Threshold)
            {
                if (state.HoverOn || InCooldown(alert.Provider, now)) continue;
                emitted.Add(alert);
                state.CooldownUntil[alert.Provider] = now + config.PerProviderCooldown;
                continue;
            }
            if (state.HoverOn || InCooldown(alert.Provider, now))
            {
                state.Held.Add(alert);
                continue;
            }
            emitted.Add(alert);
            state.CooldownUntil[alert.Provider] = now + config.PerProviderCooldown;
        }

        return emitted.OrderBy(a => a.Kind).ThenBy(a => a.Provider, StringComparer.Ordinal).ToList();
    }

    private bool InCooldown(string provider, DateTimeOffset now) =>
        state.CooldownUntil.TryGetValue(provider, out var until) && until > now;

    private void Prune(DateTimeOffset now) =>
        state.Epochs.RemoveAll(x => x.ResetsAt is DateTimeOffset at && at < now - TimeSpan.FromHours(24));
}

/// <summary>Everything the engine remembers, shaped for JSON.</summary>
public sealed class AlertState
{
    public int Schema { get; set; } = 1;
    public List<Epoch> Epochs { get; set; } = new();
    public List<WaitingEntry> Waiting { get; set; } = new();
    public List<Alert> Held { get; set; } = new();
    public Dictionary<string, DateTimeOffset> CooldownUntil { get; set; } = new();
    public bool HoverOn { get; set; }
}

public sealed class Epoch
{
    public string Key { get; set; } = "";
    public string Provider { get; set; } = "";
    public string Window { get; set; } = "";
    public DateTimeOffset? ResetsAt { get; set; }
    public double LastPct { get; set; }
    public HashSet<string> Fired { get; set; } = new();
    public bool ResetSoonFired { get; set; }
    public bool ResetDoneFired { get; set; }
    public bool Closed { get; set; }
}

public sealed class WaitingEntry
{
    public string Key { get; set; } = "";
    public string Provider { get; set; } = "";
    public string SessionId { get; set; } = "";
    public DateTimeOffset Since { get; set; }
    public DateTimeOffset? LastFired { get; set; }
}

File: macos/TokendialCore/Sources/TokendialCore/Alerts/AlertEngine.swift
import Foundation

/// The deterministic reducer specified in docs/alerts/spec.md. It owns no clock and no thread:
/// the host feeds it events and delivers what comes back. Every (provider, window) lives in an
/// epoch keyed by its reset time; a new epoch re-arms everything.
public final class AlertEngine {
    private let config: AlertConfig
    private var state: AlertState

    public init(config: AlertConfig = .default, state: AlertState? = nil) {
        self.config = config
        self.state = state ?? AlertState()
    }

    public static func load(_ json: Data, config: AlertConfig = .default) -> AlertEngine {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let state = try? decoder.decode(AlertState.self, from: json)
        return AlertEngine(config: config, state: state?.schema == 1 ? state : nil)
    }

    public func save() -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return (try? encoder.encode(state)) ?? Data()
    }

    /// Sessions the engine still believes are waiting; the host reconciles them against what it can see.
    public var waiting: [(provider: String, sessionId: String)] { state.waiting.map { ($0.provider, $0.sessionId) } }

    public func reduce(_ event: AlertEvent) -> [Alert] {
        var candidates: [Alert] = []
        switch event {
        case .usage(let at, let provider, let window, let usedPct, let resetsAt, let limited):
            candidates += applyUsage(at: at, provider: provider, window: window, usedPct: usedPct, resetsAt: resetsAt, limited: limited)
        case .session(let at, let provider, let sessionId, let activity):
            applySession(at: at, provider: provider, sessionId: sessionId, activity: activity)
        case .hover(_, let on):
            state.hoverOn = on
        case .restart(let at):
            prune(now: at)
        case .tick:
            break
        }
        candidates += timeChecks(now: event.at)
        return gate(candidates, now: event.at)
    }

    private func applyUsage(at: Date, provider: String, window: String, usedPct: Double, resetsAt: Date?, limited: Bool) -> [Alert] {
        let key = "\(provider)|\(window)"
        var index = state.epochs.firstIndex { $0.key == key }
        let existing = index.map { state.epochs[$0] }
        let fresh = existing == nil || existing!.closed || !Self.sameReset(existing!.resetsAt, resetsAt) || existing!.lastPct - usedPct > 10
        if fresh {
            state.epochs.removeAll { $0.key == key }
            state.epochs.append(Epoch(key: key, provider: provider, window: window, resetsAt: resetsAt))
            index = state.epochs.count - 1
        }
        let i = index!
        state.epochs[i].lastPct = usedPct
        state.epochs[i].resetsAt = resetsAt

        if limited || usedPct >= 100 {
            for threshold in config.thresholds { state.epochs[i].fired.insert(String(threshold)) }
            if state.epochs[i].fired.insert("limit").inserted {
                return [Alert(kind: .limit, provider: provider, window: window, resetsAt: resetsAt)]
            }
            return []
        }

        let crossed = config.thresholds.filter { Double($0) <= usedPct && !state.epochs[i].fired.contains(String($0)) }
        if crossed.isEmpty { return [] }
        for threshold in crossed { state.epochs[i].fired.insert(String(threshold)) }
        return [Alert(kind: .threshold, provider: provider, window: window, pct: crossed.max())]
    }

    private func applySession(at: Date, provider: String, sessionId: String, activity: SessionActivity) {
        let key = "\(provider)|\(sessionId)"
        if activity == .waiting {
            if !state.waiting.contains(where: { $0.key == key }) {
                state.waiting.append(WaitingEntry(key: key, provider: provider, sessionId: sessionId, since: at))
            }
            return
        }
        state.waiting.removeAll { $0.key == key }
    }

    private func timeChecks(now: Date) -> [Alert] {
        var out: [Alert] = []
        for i in state.epochs.indices {
            guard !state.epochs[i].closed, let resetsAt = state.epochs[i].resetsAt else { continue }
            let epoch = state.epochs[i]
            if !epoch.resetSoonFired && !epoch.fired.contains("limit") && resetsAt > now
                && resetsAt.timeIntervalSince(now) <= config.resetLead && epoch.lastPct >= Double(config.resetLeadMinPct) {
                state.epochs[i].resetSoonFired = true
                out.append(Alert(kind: .resetSoon, provider: epoch.provider, window: epoch.window, resetsAt: resetsAt))
            }
            if now >= resetsAt && epoch.fired.contains("limit") && !epoch.resetDoneFired {
                state.epochs[i].resetDoneFired = true
                state.epochs[i].closed = true
                out.append(Alert(kind: .resetDone, provider: epoch.provider, window: epoch.window))
            }
        }
        for i in state.waiting.indices {
            let waiting = state.waiting[i]
            let due = waiting.lastFired.map { $0.addingTimeInterval(config.waitingRepeat) } ?? waiting.since.addingTimeInterval(config.waitingDebounce)
            if now < due { continue }
            state.waiting[i].lastFired = now
            out.append(Alert(kind: .waiting, provider: waiting.provider, sessionId: waiting.sessionId))
        }
        return out
    }

    /// Hover and cooldown decide what goes out now, what waits, and what is dropped.
    private func gate(_ candidates: [Alert], now: Date) -> [Alert] {
        var emitted: [Alert] = []
        if !state.hoverOn {
            for held in state.held {
                if inCooldown(held.provider, now: now) { continue }
                state.held.removeAll { $0 == held }
                emitted.append(held)
            }
        }
        for alert in candidates {
            if alert.kind == .threshold {
                if state.hoverOn || inCooldown(alert.provider, now: now) { continue }
                emitted.append(alert)
                state.cooldownUntil[alert.provider] = now.addingTimeInterval(config.perProviderCooldown)
                continue
            }
            if state.hoverOn || inCooldown(alert.provider, now: now) {
                state.held.append(alert)
                continue
            }
            emitted.append(alert)
            state.cooldownUntil[alert.provider] = now.addingTimeInterval(config.perProviderCooldown)
        }
        return emitted.sorted { a, b in
            if a.kind.order != b.kind.order { return a.kind.order < b.kind.order }
            return a.provider.utf8.lexicographicallyPrecedes(b.provider.utf8)
        }
    }

    /// Vendors compute resets_at on the fly, so consecutive polls differ by milliseconds; a real rollover moves it by hours.
    public static let resetTolerance: TimeInterval = 60

    private static func sameReset(_ a: Date?, _ b: Date?) -> Bool {
        guard let a, let b else { return a == b }
        return abs(a.timeIntervalSince(b)) <= resetTolerance
    }

    private func inCooldown(_ provider: String, now: Date) -> Bool {
        if let until = state.cooldownUntil[provider] { return until > now }
        return false
    }

    private func prune(now: Date) {
        state.epochs.removeAll { epoch in
            if let at = epoch.resetsAt { return at < now.addingTimeInterval(-24 * 3600) }
            return false
        }
    }
}

/// Everything the engine remembers, shaped for JSON.
public struct AlertState: Codable, Sendable {
    public var schema: Int = 1
    public var epochs: [Epoch] = []
    public var waiting: [WaitingEntry] = []
    public var held: [Alert] = []
    public var cooldownUntil: [String: Date] = [:]
    public var hoverOn: Bool = false

    public init() {}
}

public struct Epoch: Codable, Sendable {
    public var key: String
    public var provider: String
    public var window: String
    public var resetsAt: Date?
    public var lastPct: Double = 0
    public var fired: Set<String> = []
    public var resetSoonFired = false
    public var resetDoneFired = false
    public var closed = false

    init(key: String, provider: String, window: String, resetsAt: Date?) {
        self.key = key
        self.provider = provider
        self.window = window
        self.resetsAt = resetsAt
    }
}

public struct WaitingEntry: Codable, Sendable {
    public var key: String
    public var provider: String
    public var sessionId: String
    public var since: Date
    public var lastFired: Date?
}
